### PR TITLE
[Account Abstraction] Add whitelisted paymasters, Add LoadOperationStorage trace, Refactor tracing, Update validation rules, Fix paymaster throttling, Fix op pool op removal

### PR DESCRIPTION
**Pull request by kristofgazso**
## Changes:
- Add option to whitelist paymasters, allowing them to access external storage
- Refactored tracing, removing UserOperationBlockTracer
- Updated validation rules, checking code hash of accessed addressed before and after instead of selfdestruct and delegatecall checking

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

### CODE REVIEW

**Review by jmederosalvarado — APPROVED** [review]
Very nice. Looks really good to me.

File: src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs [hunk 1]
```diff
@@ -94,6 +99,20 @@ public class UserOperationSimulator : IUserOperationSimulator
             UInt256? timestamp = null,
             CancellationToken cancellationToken = default)
         {
+            if (userOperation.AlreadySimulated)
+            {
+                // codehash of all accessed addresses should not change between calls
+                foreach (KeyValuePair<Address, Keccak> kv in userOperation.AddressesToCodeHashes)
+                {
+                    (Address address, Keccak expectedCodeHash) = kv;
+                    Keccak codeHash = _stateProvider.GetCodeHash(address);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs:108]: Is this access thread-safe?

File: src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs [hunk 2]
```diff
@@ -117,35 +141,52 @@ public class UserOperationSimulator : IUserOperationSimulator
...
+                userOperation.AddressesToCodeHashes = addressesToCodeHashes;
                 userOperation.AlreadySimulated = true;
             }
 
-            return ResultWrapper<Keccak>.Success(userOperation.Hash);
+            return ResultWrapper<Keccak>.Success(userOperation.RequestId!);
         }
 
-        private (bool success, UserOperationAccessList accessList, string? error) SimulateValidation(
+        private (bool success, UserOperationAccessList accessList, IDictionary<Address, Keccak> addressesToCodeHashes, string? error) SimulateValidation(
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs:151]: 4 return parameters - maybe better to put them into a class?

File: src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs [hunk 3]
```diff
@@ -117,35 +141,52 @@ public class UserOperationSimulator : IUserOperationSimulator
...
+                return (false, UserOperationAccessList.Empty, ImmutableDictionary<Address, Keccak>.Empty, error);
             }
 
-            UserOperationAccessList userOperationAccessList = new(blockTracer.AccessedStorage);
+            UserOperationAccessList userOperationAccessList = new(txTracer.AccessedStorage);
+            
+            IDictionary<Address, Keccak> addressesToCodeHashes = new Dictionary<Address, Keccak>();
+            foreach (Address accessedAddress in txTracer.AccessedAddresses)
+            {
+                addressesToCodeHashes[accessedAddress] = _stateProvider.GetCodeHash(accessedAddress);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs:186]: Is this access thread-safe?

File: src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationPool.cs [hunk 4]
```diff
@@ -108,7 +112,21 @@ private void OnBlockAdded(object? sender, BlockReplacementEventArgs e)
         {
             try
             {
-                _headBlocksChannel.Writer.TryWrite(e);
+                _headBlocksReplacementChannel.Writer.TryWrite(e);
+            }
+            catch (Exception exception)
+            {
+                if (_logger.IsError)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationPool.cs:119]: 1. What exceptions are you expecting here?
2. TryWrite probably returns a bool

File: src/Nethermind/Nethermind.Evm/VirtualMachine.cs [hunk 5]
```diff
@@ -1816,6 +1816,12 @@ void UpdateMemoryCost(in UInt256 position, in UInt256 length)
                         
                         byte[] value = _storage.Get(storageCell);
                         stack.PushBytes(value);
+                        
+                        if (_txTracer.IsTracingOpLevelStorage)
+                        {
+                            _txTracer.LoadOperationStorage(storageCell.Address, storageIndex, value);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/VirtualMachine.cs:1822]: Please evaluate all tracers for IsTracingOpLevelStorage = true

File: src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs (hunk 1 above)
Comment (kristofgazso) [reply src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs:108]: i think so, the initial simulation sets the addressesToCodeHashes and we check on the second simulation. but the second simulation can't happen until the second one does

File: src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs (hunk 3 above)
Comment (kristofgazso) [reply src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs:186]: yeah, i think so (see above)

File: src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationPool.cs (hunk 4 above)
Comment (kristofgazso) [reply src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationPool.cs:119]: i am not sure, i copied @marcindsobczak 's code from above but modified?
@marcindsobczak thoughts?

Comment (marcindsobczak) [reply src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationPool.cs:119]: We have the same in TxPool, it evolved through several changes to this form. try-catch was there before changing old method to TryWrite. I don't remember what kind of exceptions we faced when changed it to this form during TxPool crisis in August 2021. Maybe now we can do it better, but I don't know if it makes sense to change it when it works properly

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationPool.cs:119]: ok

File: src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs:151]: I would refactor this

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs (hunk 2 above)
Comment (kristofgazso) [reply src/Nethermind/Nethermind.AccountAbstraction/Executor/UserOperationSimulator.cs]: done

**Pull request merged**